File: firmware/common/hearth_httpd.cpp
#include "hearth_httpd.h"
#include <WiFi.h>
#include <Preferences.h>
#include <FS.h>

namespace hearth {

String Httpd::httpBase() const {
    return String("http://") + WiFi.localIP().toString();
}

bool Httpd::authorized(const char* path) {
    String hdr = srv_.header("Authorization");
    long ts = srv_.header("X-Hearth-Ts").toInt();
    long now = now_ ? *now_ : 0;
    return auth_ && auth_->verify(hdr, String(path), ts, now);
}

void Httpd::begin(const String& deviceId, const char* kind, const String& name,
                  const String& fw, Auth& auth, SdClip* clips,
                  const String& lanHost, const String& softap,
                  EdgeConfig* cfg, Hooks hooks, long* nowUnix) {
    deviceId_ = deviceId; kind_ = kind; name_ = name; fw_ = fw;
    auth_ = &auth; clips_ = clips; cfg_ = cfg; hooks_ = hooks;
    lanHost_ = lanHost; softap_ = softap; now_ = nowUnix;

    // We read these request headers in handlers.
    const char* wanted[] = {"Authorization", "X-Hearth-Ts"};
    srv_.collectHeaders(wanted, 2);

    srv_.on("/",          HTTP_GET,  [this]{ routeRoot(); });
    srv_.on("/status",    HTTP_GET,  [this]{ routeStatus(); });
    srv_.on("/clips",     HTTP_GET,  [this]{ routeClipsList(); });
    srv_.on("/provision", HTTP_POST, [this]{ routeProvision(); });
    srv_.on("/pair",      HTTP_POST, [this]{ routePair(); });
    srv_.on("/config",    HTTP_GET,  [this]{ routeConfigGet(); });
    srv_.on("/config",    HTTP_POST, [this]{ routeConfigPost(); });

    if (hooks_.snapshot)
        srv_.on("/snapshot", HTTP_GET, [this]{
            if (!authorized("/snapshot")) { srv_.send(401, "application/json", "{\"ok\":false}"); return; }
            if (!hooks_.snapshot(srv_)) srv_.send(503, "application/json", "{\"ok\":false}");
        });
    if (hooks_.stream)
        srv_.on("/stream", HTTP_GET, [this]{
            if (!authorized("/stream")) { srv_.send(401, "application/json", "{\"ok\":false}"); return; }
            hooks_.stream(srv_);
        });

    // /clips/<file> — match the prefix and authorize, then stream the file.
    srv_.onNotFound([this]{
        String uri = srv_.uri();
        if (uri.startsWith("/clips/")) { routeClipFile(); return; }
        srv_.send(404, "application/json", "{\"ok\":false}");
    });

    srv_.begin();
    Serial.printf("[httpd] device API up at %s/\n", httpBase().c_str());
}

void Httpd::loop() { srv_.handleClient(); }

void Httpd::routeRoot() {
    String qr = auth_->qrPayload(deviceId_, kind_, softap_, lanHost_);
    String html = "<!doctype html><meta name=viewport content='width=device-width,initial-scale=1'>";
    html += "<h3>Hearth " + String(kind_) + " — " + name_ + "</h3>";
    html += "<p><code>" + deviceId_ + "</code> fw " + fw_ + "</p>";
    html += "<p>Pair payload (QR):</p><pre style='white-space:pre-wrap'>" + qr + "</pre>";
    html += "<p><a href=/status>/status</a>";
    if (hooks_.snapshot) html += " &middot; /snapshot";
    if (hooks_.stream)   html += " &middot; /stream";
    if (clips_)          html += " &middot; <a href=/clips>/clips</a>";
    html += "</p>";
    srv_.send(200, "text/html", html);
}

void Httpd::routeStatus() {
    String j = "{";
    j += "\"device_id\":\"" + deviceId_ + "\",";
    j += "\"kind\":\""      + String(kind_) + "\",";
    j += "\"name\":\""      + name_ + "\",";
    j += "\"fw\":\""        + fw_ + "\",";
    j += "\"ip\":\""        + WiFi.localIP().toString() + "\",";
    j += "\"rssi\":"        + String((long)WiFi.RSSI()) + ",";
    j += "\"lan_host\":\""  + lanHost_ + "\"";
    if (hooks_.statusExtra) { String x = hooks_.statusExtra(); if (x.length()) j += "," + x; }
    j += "}";
    srv_.send(200, "application/json", j);
}

void Httpd::routeClipsList() {
    if (!authorized("/clips")) { srv_.send(401, "application/json", "{\"ok\":false}"); return; }
    if (!clips_) { srv_.send(200, "application/json", "[]"); return; }
    srv_.send(200, "application/json", clips_->listJson());
}

void Httpd::routeClipFile() {
    if (!authorized(srv_.uri().c_str())) { srv_.send(401, "application/json", "{\"ok\":false}"); return; }
    if (!clips_ || !clips_->fs()) { srv_.send(404, "application/json", "{\"ok\":false}"); return; }
    String file = srv_.uri().substring(strlen("/clips/"));
    if (file.indexOf("..") >= 0) { srv_.send(400, "application/json", "{\"ok\":false}"); return; }
    String path = String(clips_->dir()) + "/" + file;
    File f = clips_->fs()->open(path, FILE_READ);
    if (!f) { srv_.send(404, "application/json", "{\"ok\":false}"); return; }
    srv_.streamFile(f, "video/x-motion-jpeg");
    f.close();
}

void Httpd::routeProvision() {
    String ssid = srv_.arg("ssid"), pass = srv_.arg("pass");
    if (!ssid.length()) { srv_.send(400, "application/json", "{\"ok\":false}"); return; }
    srv_.send(200, "application/json", "{\"ok\":true}");
    // Persist via the same NVS namespace/keys hearth_wifi uses, then reboot to
    // re-run STA join with the new creds.
    Preferences p; p.begin("hearth", false);
    p.putString("wifi_ssid", ssid); p.putString("wifi_pass", pass); p.end();
    delay(200); ESP.restart();
}

void Httpd::routePair() {
    // §6: phone POSTs the QR `key` to prove possession; we return a device token.
    // The token here is HMAC(key, "/pair"+ts) — the phone caches {device_id,key}
    // and self-derives per-request bearers, so /pair just confirms the key works.
    long ts = srv_.header("X-Hearth-Ts").toInt();
    if (!authorized("/pair")) { srv_.send(401, "application/json", "{\"ok\":false}"); return; }
    String tok = auth_->sign("/pair", ts);
    String j = "{\"ok\":true,\"device_id\":\"" + deviceId_ + "\",\"token\":\"" + tok + "\"}";
    srv_.send(200, "application/json", j);
}

void Httpd::routeConfigGet() {
    if (!cfg_) { srv_.send(200, "application/json", "{}"); return; }
    String j = "{";
    j += "\"person_threshold\":" + String(cfg_->person_threshold, 2) + ",";
    j += "\"retention_days\":"   + String(cfg_->retention_days) + ",";
    j += "\"face\":" + String(cfg_->face ? "true" : "false") + "}";
    srv_.send(200, "application/json", j);
}

void Httpd::routeConfigPost() {
    if (!authorized("/config")) { srv_.send(401, "application/json", "{\"ok\":false}"); return; }
    if (!cfg_) { srv_.send(200, "application/json", "{\"ok\":true}"); return; }
    if (srv_.hasArg("person_threshold")) cfg_->person_threshold = srv_.arg("person_threshold").toFloat();
    if (srv_.hasArg("retention_days"))   cfg_->retention_days   = srv_.arg("retention_days").toInt();
    if (srv_.hasArg("face"))             cfg_->face = (srv_.arg("face") == "true" || srv_.arg("face") == "1");
    if (hooks_.onConfig) hooks_.onConfig(*cfg_);
    srv_.send(200, "application/json", "{\"ok\":true}");
}

} // namespace hearth

File: firmware/cam-k230/hearth_httpd.py
# hearth_httpd — device HTTP API (FABRIC.md §6) for the K230, a tiny socket
# server run in a background thread. Endpoints: / /status /snapshot /clips
# /clips/<f> /config (GET/POST) /pair. Media + clips require the HMAC bearer.
import socket, ujson, _thread, os

class Httpd:
    def __init__(self, device_id, kind, name, fw, auth, clips_dir,
                 lan_host, softap, cfg, snapshot_fn, now_fn):
        self.device_id = device_id; self.kind = kind; self.name = name; self.fw = fw
        self.auth = auth; self.clips_dir = clips_dir
        self.lan_host = lan_host; self.softap = softap
        self.cfg = cfg                      # dict: person_threshold/retention_days/face
        self.snapshot_fn = snapshot_fn      # () -> jpeg bytes or None
        self.now_fn = now_fn                # () -> unix seconds

    def start(self):
        _thread.start_new_thread(self._serve, ())

    def http_base(self, ip):
        self._ip = ip
        return "http://" + ip

    def _authorized(self, headers, path):
        return self.auth.verify(headers.get("authorization", ""), path,
                                headers.get("x-hearth-ts", "0"), self.now_fn())

    def _serve(self):
        s = socket.socket(); s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        s.bind(("0.0.0.0", 80)); s.listen(2)
        while True:
            try:
                cl, _ = s.accept()
                self._handle(cl)
            except Exception as e:
                print("[httpd]", e)

    def _handle(self, cl):
        req = cl.recv(2048).decode("utf-8", "ignore")
        if not req:
            cl.close(); return
        line = req.split("\r\n", 1)[0]
        parts = line.split(" ")
        method, path = (parts + ["", "/"])[:2]
        path = path.split("?", 1)[0]
        headers = {}
        for h in req.split("\r\n")[1:]:
            if ":" in h:
                k, v = h.split(":", 1); headers[k.strip().lower()] = v.strip()
        body = req.split("\r\n\r\n", 1)[-1]

        if path == "/" and method == "GET":
            qr = self.auth.qr_payload(self.device_id, self.kind, self.softap, self.lan_host)
            self._send(cl, 200, "text/html",
                       "<h3>Hearth %s — %s</h3><pre>%s</pre>" % (self.kind, self.name, qr))
        elif path == "/status" and method == "GET":
            self._json(cl, 200, {"device_id": self.device_id, "kind": self.kind,
                                 "name": self.name, "fw": self.fw,
                                 "lan_host": self.lan_host,
                                 "capabilities": {"stream": True, "snapshot": True,
                                                  "clips": True, "person_detect": "reliable",
                                                  "sd": True}})
        elif path == "/snapshot" and method == "GET":
            if not self._authorized(headers, path): return self._json(cl, 401, {"ok": False})
            jpg = self.snapshot_fn()
            if not jpg: return self._json(cl, 503, {"ok": False})
            self._send_bytes(cl, 200, "image/jpeg", jpg)
        elif path == "/clips" and method == "GET":
            if not self._authorized(headers, path): return self._json(cl, 401, {"ok": False})
            self._send(cl, 200, "application/json", self._clips_json())
        elif path.startswith("/clips/") and method == "GET":
            if not self._authorized(headers, path): return self._json(cl, 401, {"ok": False})
            self._send_file(cl, path[len("/clips/"):])
        elif path == "/config" and method == "GET":
            self._json(cl, 200, self.cfg)
        elif path == "/config" and method == "POST":
            if not self._authorized(headers, path): return self._json(cl, 401, {"ok": False})
            try: self.cfg.update(ujson.loads(body))
            except Exception: pass
            self._json(cl, 200, {"ok": True})
        elif path == "/pair" and method == "POST":
            if not self._authorized(headers, path): return self._json(cl, 401, {"ok": False})
            ts = headers.get("x-hearth-ts", "0")
            self._json(cl, 200, {"ok": True, "device_id": self.device_id,
                                 "token": self.auth.sign("/pair", ts)})
        else:
            self._json(cl, 404, {"ok": False})
        cl.close()

    def _clips_json(self):
        out = []
        try:
            for f in os.listdir(self.clips_dir):
                try: sz = os.stat(self.clips_dir + "/" + f)[6]
                except OSError: sz = 0
                out.append({"file": f, "ts": int(f.split(".")[0]) if f[0].isdigit() else 0, "size": sz})
        except OSError:
            pass
        return ujson.dumps(out)

    def _send_file(self, cl, fname):
        if ".." in fname: return self._json(cl, 400, {"ok": False})
        try:
            f = open(self.clips_dir + "/" + fname, "rb")
        except OSError:
            return self._json(cl, 404, {"ok": False})
        cl.send(b"HTTP/1.0 200 OK\r\nContent-Type: video/mp4\r\n\r\n")
        while True:
            chunk = f.read(1024)
            if not chunk: break
            cl.send(chunk)
        f.close()

    def _send(self, cl, code, ctype, body):
        cl.send(("HTTP/1.0 %d OK\r\nContent-Type: %s\r\n\r\n" % (code, ctype)).encode())
        cl.send(body.encode() if isinstance(body, str) else body)

    def _send_bytes(self, cl, code, ctype, data):
        cl.send(("HTTP/1.0 %d OK\r\nContent-Type: %s\r\nContent-Length: %d\r\n\r\n"
                 % (code, ctype, len(data))).encode())
        cl.send(data)

    def _json(self, cl, code, obj):
        self._send(cl, code, "application/json", ujson.dumps(obj))
